Vungle/VungleRouter.m

```diff
@@ -287,36 +288,38 @@ - (void)requestBannerAdWithPlacementID:(NSString *)placementID
-        [self.bannerDelegates addObject:dictionary];
-    } else {
-        [dictionary setObject:delegate forKey:kVungleBannerDelegateKey];
-        [dictionary setObject:[NSNumber numberWithInt:BannerRouterDelegateStateRequesting] forKey:kVungleBannerDelegateStateKey];
-        [self.bannerDelegates addObject:dictionary];
-        
-        NSError *error = nil;
-        if (CGSizeEqualToSize(size, kVNGMRECSize)) {
-            if ([[VungleSDK sharedSDK] loadPlacementWithID:placementID error:&error]) {
-                MPLogInfo(@"Vungle: Start to load an ad for Placement ID :%@", placementID);
-            } else {
-                [self requestBannerAdFailedWithError:error
-                                         placementID:placementID
-                                            delegate:delegate];
-            }
+    @synchronized (self ) {
+        NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
+        if ([self isBannerAdAvailableForPlacementId:placementID size:size]) {
+            [delegate vungleAdDidLoad];
+
+            [dictionary setObject:delegate forKey:kVungleBannerDelegateKey];
+            [dictionary setObject:[NSNumber numberWithInt:BannerRouterDelegateStateCached] forKey:kVungleBannerDelegateStateKey];
+            [self.bannerDelegates addObject:dictionary];
         } else {
-            if ([[VungleSDK sharedSDK] loadPlacementWithID:placementID withSize:[self getVungleBannerAdSizeType:size] error:&error]) {
-                MPLogInfo(@"Vungle: Start to load an ad for Placement ID :%@", placementID);
-            } else {
-                if ((error) && (error.code != VungleSDKResetPlacementForDifferentAdSize)) {
+            [dictionary setObject:delegate forKey:kVungleBannerDelegateKey];
+            [dictionary setObject:[NSNumber numberWithInt:BannerRouterDelegateStateRequesting] forKey:kVungleBannerDelegateStateKey];
+            [self.bannerDelegates addObject:dictionary];
+
+            NSError *error = nil;
+            if (CGSizeEqualToSize(size, kVNGMRECSize)) {
+                if ([[VungleSDK sharedSDK] loadPlacementWithID:placementID error:&error]) {
+                    MPLogInfo(@"Vungle: Start to load an ad for Placement ID :%@", placementID);
+                } else {
                     [self requestBannerAdFailedWithError:error
                                              placementID:placementID
                                                 delegate:delegate];
                 }
+            } else {
+                if ([[VungleSDK sharedSDK] loadPlacementWithID:placementID withSize:[self getVungleBannerAdSizeType:size] error:&error]) {
+                    MPLogInfo(@"Vungle: Start to load an ad for Placement ID :%@", placementID);
+                } else {
+                    if ((error) && (error.code != VungleSDKResetPlacementForDifferentAdSize)) {
```

Comment:
@YueVungle 
I believe we got one concern from AdMob side for this line. This was my bad API design, but SDK will return "error" and "NO" for this case (ad cached but requested with different size). As you see in SDK code (`VunglePlacementsCoordinator`), SDK will return error and NO but it continue to request an ad with new size.
So, you don't need to catch this case and send error callback (equestBannerAdFailedWithError) as below.

cc @shiv-vungle @crbinsf

Reply:
@ashinagawa Hi Aki, yes, I have checked the method `loadPlacementWithID:withSize:error:` in `VunglePlacementsCoordinator` class, SDK will return error and NO but it continues to request an ad with new size.  We don't need to catch this case and send error callback. So we will keep this piece of logic and not remove it, right?  Thanks a lot. ^_^

Reply:
We DO need to remove this. We don't want to send `requestBannerAdFailedWithError` here since request is continued and adapter will get "playabilityUpdate" later. So please remove line 316-322 code.

Reply:
We might need to remove the logic in SDK in future but for now, we need to remove above section in adapter.

Reply:
@ashinagawa Hi Aki, I have a concern about removing line 316-322 code, if `[[VungleSDK sharedSDK] loadPlacementWithID:placementID withSize:[self getVungleBannerAdSizeType:size] error:&error])` return NO because of other errors (not the `VungleSDKResetPlacementForDifferentAdSize` error), we should send `requestBannerAdFailedWithError` here.